File: vue/src/server/url_into_sheet.py
import json
import time
import socket
import gspread
import requests
from oauth2client.service_account import ServiceAccountCredentials
import os


def get_sheet(spreadsheet_id, sheet_name):
    # use creds to create a client to interact with the Google Drive API
    scope = ['https://spreadsheets.google.com/feeds', 'https://www.googleapis.com/auth/drive']

    # relative to script dir
    script_dir = os.path.dirname(__file__)
    secrets_file = os.path.join(script_dir, '../../../secrets/googlekey.json')
    creds = ServiceAccountCredentials.from_json_keyfile_name(secrets_file, scope)
    client = gspread.authorize(creds)
    try:
        sheet = client.open_by_key(spreadsheet_id).worksheet(sheet_name)
    except gspread.exceptions.WorksheetNotFound:
        client.open_by_key(spreadsheet_id).add_worksheet(sheet_name, 100, 100)
        sheet = client.open_by_key(spreadsheet_id).worksheet(sheet_name)
    return sheet


def edit_row(spreadsheet_id, sheet_name):
    sheet = get_sheet(spreadsheet_id, sheet_name)
    # get current ip using dev.get_ip()
    url = get_ngrok_url()
    # Update the specified row with the given values
    sheet.update('A1', "IP")
    sheet.update('B1', url)
    sheet.update('A2', "last update")
    sheet.update('B2', time.ctime())


def get_ngrok_url():
    try:
        res = requests.get("http://localhost:4040/api/tunnels")
        # Get the JSON data from the response
        res_data = res.json()
        # Get the public URL of the first tunnel
        public_url = res_data["tunnels"][0]["public_url"]
    except Exception as e:
        return str(e)
    return public_url


def write_url_to_google_sheet():
    with open('secrets/googlesheet.json') as f:
        spreadsheet_id = json.load(f)['id']
    hostname = socket.gethostname()
    edit_row(spreadsheet_id, hostname)


if __name__ == '__main__':
    write_url_to_google_sheet()

File: vue/src/server/url_into_sheet.js
const { google } = require("googleapis");
const path = require('path');

const currentDirrectory = `${path.dirname(require.main.filename)}`

const auth = new google.auth.GoogleAuth({
  keyFile: `${currentDirrectory}/../../../secrets/googlekey.json`,
  scopes: [
    "https://spreadsheets.google.com/feeds",
    "https://www.googleapis.com/auth/drive",
  ],
});
const sheets = google.sheets({version: "v4", auth});

async function createSheetWithTitle(spreadsheetId, title) {
  const resource = {
    requests: [
      {
        addSheet: {
          properties: {
            title: title
          }
        }
      }
    ]
  };
  const response = await sheets.spreadsheets.batchUpdate({
    spreadsheetId,
    resource,
  });
  console.log(response.data);
  sheets.spreadsheets.get()
  return response.data
}

async function deleteSheet(spreadsheetId, sheetId) {
  const resource = {
    requests: [
      {
        deleteSheet: {
          sheetId,
        },
      },
    ],
  };
  const response = await sheets.spreadsheets.batchUpdate({
    spreadsheetId,
    resource,
  });
  console.log(response.data);
}


async function createSheetIfNotExists(spreadsheetId, title) {
  const spreadsheet = await sheets.spreadsheets.get({spreadsheetId: spreadsheetId});
  var sheet = spreadsheet.data.sheets.find(s => s.properties.title === title);
  if (sheet !== undefined) {
    return sheet
  }
  return createSheetWithTitle(spreadsheetId, title)
}

async function updateCells(spreadsheetId, range, values) {
  const resource = {
    values,
  };
  try {
    const response = await sheets.spreadsheets.values.update({
      spreadsheetId,
      range,
      valueInputOption: 'USER_ENTERED',
      resource,
    });
    console.log(response.data);
  } catch (err) {
    console.error(err);
  }
}

async function updateNamedSheet(spreadsheetId, sheetName, url, time) {
  const range = `${sheetName}!A1:B2`;
  const values = [
    ['IP', url],
    ['last update', time],
  ];
  await updateCells(spreadsheetId, range, values)
}

async function findRangeByValue(spreadsheetId, sheetName, value) {
  const range = `${sheetName}!A1:C`;
  const response = await sheets.spreadsheets.values.get({
    spreadsheetId,
    range,
  });
  const rows = response.data.values;
  var targetRow = response.data.values ? response.data.values.length + 1 : 1;
  for (let i = 0; i < rows.length; i++) {
    if (rows[i].indexOf(value) !== -1) {
      targetRow = i + 1;
      break;
    }
  }
  return `${sheetName}!A${targetRow}:C${targetRow}`;
}

async function updateMainSheet(spreadsheetId, sheetName, url, time) {
  const range = await findRangeByValue(spreadsheetId, 'main', sheetName);
  const values = [
    [sheetName, time, url],
  ];
  await updateCells(spreadsheetId, range, values)
}

async function saveUrlToSpreadsheet(spreadsheetId, sheetName, url) {
  const time = new Date().toString();
  await createSheetIfNotExists(spreadsheetId, sheetName);
  await updateNamedSheet(spreadsheetId, sheetName, url, time)
  await updateMainSheet(spreadsheetId, sheetName, url, time)
}

module.exports = {
  writeUrlToGoogleSheet: saveUrlToSpreadsheet
};
